go_lib/deckhouse-config/validator.go

```diff
@@ -168,18 +172,32 @@ func (c *ConfigValidator) Validate(cfg *d8cfg_v1alpha1.ModuleConfig) ValidationR
 // validateSettings uses ValuesValidator from ModuleManager instance to validate spec.settings.
 // cfgName arg is a kebab-cased name of the ModuleConfig resource.
 // cfgSettings is a content of spec.settings.
-// (Note: cfgValues map is a map with 'plain values', i.e. without camelCased module name as a root key).
+// (Note: cfgSettings map is a map with 'plain values', i.e. without camelCased module name as a root key).
 func (c *ConfigValidator) validateSettings(cfgName string, cfgSettings map[string]interface{}) error {
 	// Ignore empty validator.
 	if c.valuesValidator == nil {
 		return nil
 	}
 
 	valuesKey := valuesKeyFromObjectName(cfgName)
-	values := map[string]interface{}{
-		valuesKey: cfgSettings,
+	schemaType := validation.ModuleSchema
+	if cfgName == "global" {
+		schemaType = validation.GlobalSchema
 	}
 
+	// Instantiate defaults from the OpenAPI schema.
+	defaultSettings := make(map[string]interface{})
+	s := c.valuesValidator.GetSchema(schemaType, validation.ConfigValuesSchema, valuesKey)
+	if s != nil {
+		validation.ApplyDefaults(defaultSettings, s)
+	}
+
+	// Merge defaults with passed settings as addon-operator will do.
+	values := utils.MergeValues(
+		utils.Values{valuesKey: defaultSettings},
+		utils.Values{valuesKey: cfgSettings},
+	)
+
 	if cfgName == "global" {
 		return c.valuesValidator.ValidateGlobalConfigValues(values)
```

Comment:
Why do we treat global values and module values separately? It's fine that we load them from different places. But it looks like unnecessary interface bloat to me.

It's more for future cosideration than for addressing in current PR.

Reply:
Yeah, it should be addressed in the Addon-operator. I've choose no to wrap these methods and use them as-is. See [validator.go](https://github.com/flant/addon-operator/blob/main/pkg/values/validation/validator.go#L25-L39)